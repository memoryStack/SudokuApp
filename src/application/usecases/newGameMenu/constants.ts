// TODO: how to compose enums so that's it's easier to edit or re-organize them
// TODO: these names are confusing, fix these
export enum AUTO_GENERATED_NEW_GAME_IDS {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
}

export enum NEW_GAME_IDS {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
    CUSTOMIZE_PUZZLE = 'CUSTOMIZE_PUZZLE',
}

export enum START_GAME_MENU_ITEMS_IDS {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
    CUSTOMIZE_PUZZLE = 'CUSTOMIZE_PUZZLE',
    RESUME = 'RESUME'
}
