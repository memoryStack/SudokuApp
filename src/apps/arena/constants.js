import { roundToNearestPixel } from '../../utils/util'

export const BOARD_AXES_VALUES = {
    Y_AXIS: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
    X_AXIS: ['1', '2', '3', '4', '5', '6', '7', '8', '9'],
}

export const BOARD_CELLS_COUNT = 81
export const MAX_INSTANCES_OF_NUMBER = 9
export const CELLS_IN_HOUSE = 9
export const NUMBERS_IN_HOUSE = 9
export const HOUSES_COUNT = 9
export const BOARD_TO_WINDOW_WIDTH_RATIO = 0.94
export const THICK_BORDER_COUNT = 2
export const THIN_BORDER_COUNT = 8
export const BLOCKS_COUNT_IN_ROW = 3

export const GRID_TRAVERSALS = {
    ROW: 'ROW',
    COL: 'COL',
    BLOCK: 'BLOCK',
}

export const STATIC_BOARD_ELEMENTS_DIMENSIONS = {
    THICK_BORDER_WIDTH: roundToNearestPixel(3),
    THIN_BORDER_WIDTH: roundToNearestPixel(1),
    AXIS_WIDTH: roundToNearestPixel(20),
}

export const DEEPLINK_PUZZLE_URL_ERRORS = {
    EMPTY_URL: 'Invaild Url',
    INCOMPLETE_NUMBERS: `Puzzle doesn't have ${BOARD_CELLS_COUNT} numbers`,
    INVALID_CHARACTERS: 'Puzzle has invalid characters other than 0-9',
}

export const PUZZLE_SOLUTION_TYPES = {
    NO_SOLUTION: 'NO_SOLUTION',
    UNIQUE_SOLUTION: 'UNIQUE_SOLUTION',
    MULTIPLE_SOLUTIONS: 'MULTIPLE_SOLUTIONS',
}

export const BOARD_MOVES_TYPES = {
    ADD: 'ADD',
    REMOVE: 'REMOVE',
}

export const BOARD_GRID_BORDERS_DIRECTION = {
    VERTICAL: 'vertical',
    HORIZONTAL: 'horizontal',
}

export const ARENA_PAGE_TEST_ID = 'arena_test_id'

export const GAME_OVER_CARD_OVERLAY_TEST_ID = 'GAME_OVER_CARD_OVERLAY_TEST_ID'
