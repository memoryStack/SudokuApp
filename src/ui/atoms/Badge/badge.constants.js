export const BADGE_TYPE = {
    SMALL: 'sm',
    LARGE: 'lg',
}
