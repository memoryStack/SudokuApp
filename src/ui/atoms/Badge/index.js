export { default } from './Badge'
export * from './badge.constants'
