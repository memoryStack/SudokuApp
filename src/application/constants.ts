import { AUTO_GENERATED_NEW_GAME_IDS } from "./usecases/newGameMenu/constants"

export enum BOARD_MOVES_TYPES {
    ADD = 'ADD',
    REMOVE = 'REMOVE',
}

export enum PENCIL_STATE {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
}

export enum GAME_STATE {
    ACTIVE = 'ACTIVE',
    INACTIVE = 'INACTIVE',
    DISPLAY_HINT = 'DISPLAY_HINT',
    GAME_SELECT = 'GAME_SELECT',
    OVER_SOLVED = 'SOLVED',
    OVER_UNSOLVED = 'UNSOLVED',
}

export enum LEVEL_DIFFICULTIES {
    EASY = 'EASY',
    MEDIUM = 'MEDIUM',
    HARD = 'HARD',
    EXPERT = 'EXPERT',
}

export const GAME_LEVEL_VS_CLUES = {
    [AUTO_GENERATED_NEW_GAME_IDS.EASY]: 36,
    [AUTO_GENERATED_NEW_GAME_IDS.MEDIUM]: 32,
    [AUTO_GENERATED_NEW_GAME_IDS.HARD]: 28,
    [AUTO_GENERATED_NEW_GAME_IDS.EXPERT]: 24,
}
