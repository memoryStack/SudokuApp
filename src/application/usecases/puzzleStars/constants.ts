import { NEW_GAME_IDS } from "../newGameMenu/constants"

const TIME_LIMIT_FOR_ONE_STAR = 100000000

// this timeLimit is in minutes
export const PUZZLE_STARS_DISTRIBUTION = {
    [NEW_GAME_IDS.EASY]: [{ stars: 3, timeLimit: 20 }, { stars: 2, timeLimit: 30 }, { stars: 1, timeLimit: TIME_LIMIT_FOR_ONE_STAR }],
    [NEW_GAME_IDS.MEDIUM]: [{ stars: 3, timeLimit: 30 }, { stars: 2, timeLimit: 40 }, { stars: 1, timeLimit: TIME_LIMIT_FOR_ONE_STAR }],
    [NEW_GAME_IDS.HARD]: [{ stars: 3, timeLimit: 40 }, { stars: 2, timeLimit: 50 }, { stars: 1, timeLimit: TIME_LIMIT_FOR_ONE_STAR }],
    [NEW_GAME_IDS.EXPERT]: [{ stars: 3, timeLimit: 50 }, { stars: 2, timeLimit: 60 }, { stars: 1, timeLimit: TIME_LIMIT_FOR_ONE_STAR }]
}
