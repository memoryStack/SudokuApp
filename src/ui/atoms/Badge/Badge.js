import React from 'react'

import { ViewPropTypes } from 'react-native'

import PropTypes from 'prop-types'

import { useStyles } from '@utils/customHooks/useStyles'

import Text from '../Text'

import { BADGE_TEST_ID, BADGE_TYPE } from './badge.constants'
import { getStyles } from './badge.styles'

/*
    How to use this properly
        <View style={{  // this view will bind the icon and the badge together
            marginTop: 40,
            borderWidth: 2,
            borderColor: 'black',
        }}>
            // this view will decide the layout sizing, it will be an Icon/Image
            // or some other view which will have the badge over it
            <View style={{ height: 32,width: 32 }} />

            // this is absolutely positioned anyway
            <Badge label="99999+" />
        </View>

        Note: the view which is binding Icon/Image should be not limited for space
            else the Badge will shrink and won't be able to expand properly
        TODO: this demands a research if an absolutely positioned element can be
        intrinsicly(without height/width set) wider than it's containing block.
*/

const Badge = ({
    type,
    label,
    styles: stylesProp,
}) => {
    const styles = useStyles(getStyles, { type })

    const renderLabel = () => (type === BADGE_TYPE.SMALL ? '' : label)

    return (
        <Text
            testID={BADGE_TEST_ID}
            style={[styles.label, styles.container, stylesProp]}
        >
            {renderLabel()}
        </Text>
    )
}

export default React.memo(Badge)

Badge.propTypes = {
    type: PropTypes.string,
    label: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
    ]),
    styles: ViewPropTypes.style,
}

Badge.defaultProps = {
    type: BADGE_TYPE.LARGE,
    label: '',
    styles: {},
}
