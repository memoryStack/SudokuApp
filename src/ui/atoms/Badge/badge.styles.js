import { StyleSheet } from 'react-native'

import _get from '@lodash/get'

import { BADGE_TYPE } from './badge.constants'

export const getStyles = ({ type }, theme) => {
    const size = _get(theme, ['badge', type, 'container', 'layout', 'size'])

    const sizeProps = {
        height: size,
        ...type === BADGE_TYPE.LARGE && { minWidth: size },
        ...type === BADGE_TYPE.SMALL && { width: size },
    }

    // NOTE: after assuming that all the icons/images with badges on them
    //      will be of 32*32 need to standardize the icon sizes
    const placementProps = {
        top: type === BADGE_TYPE.SMALL ? 0 : -4,
        ...type === BADGE_TYPE.SMALL && { right: 0 },
        ...type === BADGE_TYPE.LARGE && { left: 16 },
    }

    return StyleSheet.create({
        container: {
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            position: 'absolute',
            ...placementProps,
            backgroundColor: _get(theme, ['badge', type, 'container', 'color']),
            ..._get(theme, ['badge', type, 'container', 'layout', 'shape']),
            padding: _get(theme, ['badge', type, 'container', 'layout', 'padding']),
            ...sizeProps,
        },
        label: {
            ...type === BADGE_TYPE.LARGE && {
                color: _get(theme, ['badge', type, 'label', 'color']),
                ..._get(theme, ['badge', type, 'label', 'font']),
            },
        },
    })
}
