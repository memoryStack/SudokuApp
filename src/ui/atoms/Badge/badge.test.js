import * as React from 'react'

import { render, screen } from '@utils/testing/testingLibrary'

import Badge from './Badge'
import { BADGE_TYPE, BADGE_TEST_ID } from './badge.constants'

test('snapshot for small badge', () => {
    const tree = render(<Badge type={BADGE_TYPE.SMALL} />).toJSON()

    expect(tree).toMatchSnapshot()
})

test('snapshot for large badge', () => {
    const tree = render(<Badge type={BADGE_TYPE.LARGE} label={34} />).toJSON()

    expect(tree).toMatchSnapshot()
})

describe('badge functionality', () => {
    test('for small badges label is ignored', () => {
        render(<Badge type={BADGE_TYPE.SMALL} label={34789} />)

        expect(screen.getByTestId(BADGE_TEST_ID)).toBeEmptyElement()
    })

    test('Styles Overrides', () => {
        const customStyles = { backgroundColor: 'green' }
        render(
            <Badge
                type={BADGE_TYPE.SMALL}
                styles={customStyles} // TODO: sync these prop names
            />,
        )

        expect(screen.getByTestId(BADGE_TEST_ID)).toHaveStyle(customStyles)
    })
})
