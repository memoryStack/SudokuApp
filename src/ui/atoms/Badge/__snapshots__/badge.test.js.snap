// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for large badge 1`] = `
<Text
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#AF251D",
      "borderRadius": 1000,
      "color": "#FFFFFF",
      "display": "flex",
      "fontFamily": "OpenSans-Medium",
      "fontSize": 11,
      "fontWeight": "500",
      "height": 16,
      "includeFontPadding": false,
      "justifyContent": "center",
      "left": 16,
      "letterSpacing": 0.5,
      "lineHeight": 16,
      "minWidth": 16,
      "padding": 0,
      "position": "absolute",
      "textAlign": "center",
      "textAlignVertical": "center",
      "top": -4,
    }
  }
  testID="badge_test_id"
>
  34
</Text>
`;

exports[`snapshot for small badge 1`] = `
<Text
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#AF251D",
      "borderRadius": 1000,
      "display": "flex",
      "fontFamily": "OpenSans",
      "fontSize": 16,
      "fontWeight": "400",
      "height": 6,
      "includeFontPadding": false,
      "justifyContent": "center",
      "letterSpacing": 0.5,
      "lineHeight": 24,
      "padding": undefined,
      "position": "absolute",
      "right": 0,
      "textAlign": "center",
      "textAlignVertical": "center",
      "top": 0,
      "width": 6,
    }
  }
  testID="badge_test_id"
>
  
</Text>
`;
